rpc/src/responses/chains/blocks/invalid_blocks_in_chain.rs

```diff
@@ -0,0 +1,124 @@
+use crate::errors::ParseError;
+use crate::responses::{json_array, Response};
+use crate::types::Unistring;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::collections::HashMap;
+
+#[derive(Serialize, Deserialize)]
+pub struct InvalidBlockError {
+	pub kind: String,
+	pub id: String,
+	pub invalid_block: Unistring,
+	pub error: String,
+	#[serde(flatten)]
+	pub extra_error_info: HashMap<String, Value>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct InvalidBlock {
+	pub block: Unistring,
+	pub level: i32,
+	pub errors: InvalidBlockError,
+}
+
+pub struct InvalidBlocksInChainResponse {
+	pub invalid_blocks: json_array::JsonArray<InvalidBlock>,
+}
+
+impl Response for InvalidBlocksInChainResponse {
+	/// Parses a response string in the form
+	/// `"[{ "block": "alpha_numeric_block_id_string",
+	///      "level": integer ∈ [-2^31-1, 2^31],
+	///      "errors": $error }, ...]"` into a
+	/// [`InvalidBlocksInChainResponse`](Self).
+	fn from_response_str(response: &str) -> Result<Self, ParseError> {
+		let invalid_blocks = json_array::JsonArray::from_str(response)?;
+
+		Ok(Self { invalid_blocks })
+	}
+}
+
+#[cfg(test)]
+mod test {
+	use super::*;
+
+	#[test]
+	fn get_invalid_blocks_in_chain_from_response_empty_fails() {
+		let mock_response = "";
+
+		let blocks_response = InvalidBlocksInChainResponse::from_response_str(mock_response);
+		assert!(blocks_response.is_err());
+	}
+
+	#[test]
+	fn get_invalid_blocks_in_chain_from_response_ok() {
+		let mock_block = "blockId1";
+		let mock_level = 1;
+
+		let mock_kind = "permanent";
+		let mock_id = "validator.invalid_block";
+		let mock_invalid_block_hash = "blockId1";
+		let mock_error = "cannot_parse_operation";
+		let mock_operation = "operationHash1";
+
+		let mock_error_response = format!(
+			r#"{{
+				"kind":"{}",
+				"id":"{}",
+				"invalid_block":"{}",
+				"error": "{}",
+				"operation": "{}"
+			}}"#,
+			mock_kind, mock_id, mock_invalid_block_hash, mock_error, mock_operation
+		);
+
+		let mock_response = format!(
+			r#"[{{
+				"block":"{}",
+				"level":{},
+				"errors": {}
+			}}]"#,
+			mock_block, mock_level, mock_error_response
+		);
```

Comment:
On second thought, this is mostly unnecessary string manipulation and it feels like a massive sin to have so much ungrouped but tightly coupled data.

A much cleaner solution would be to just create an `InvalidBlockError` and then make it `impl Display` or add a `to_string()` method that returns this. 

In this way we could have better equality checks, no string manipulations at all, and have _really_ nice debug messages down the line since `Display` will be satisfied.

Reply:
I agree that it is a bit weird to have so much ungrouped but tightly coupled data as you said but this is solely for the purpose of being able to test each value against `invalid_blocks_response`.

I could implement `Display` or add a `to_string()` method but I'm not totally sure if this would be worth it because this would only be used in testing (since we would get a full response from the RPC). So since it would only be useful in testing it might be better to not introduce any unnecessary complexity and keep it simple with `format!` so future collaborators don't have to refer to structs and understand what the testing is doing instantly.